### PR TITLE
Tommy Takahashi

### PR DESCRIPTION
**Pull request by ttakahashi1591**
I believe that I did okay on this IC, but I know I did not get far enough to pass. Some things I want to practice more is getting my test data set-up more easily and I definitely need to keep a close eye on my spelling!!

### CODE REVIEW

File: app/controllers/customers_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class CustomersController < ApplicationController
+  def index
+
+  end
+  
+  def show
+    @customer = Customer.find(params[:id])
+
+    if request.post? && params[:customer].present? && params[:customer][:item_id].present?
```
Comment (juliet-e) [inline app/controllers/customers_controller.rb:9]: While it is okay for one controller action to receipt multiple types of requests, this usually only happens when those requests are really similar. In this case, I wouldn't have your form redirect back to this show action. It's not the most RESTful option because the `show` action is generally just a data displaying action (i.e. grab some data from the db and hand it to the view) but what you're doing here is `creating` something -- a relationship between a customer and an item. 

So instead of a `show` action, we want a `create` action!

Where to put the `create` action is a different question. If we put it in this same CustomersController, it suggests that we are creating a new customer. But we're not creating a new customer but just a new relationship between existing customers and items. So, what is conventional here is to create a new *joins* controller and call it CustomerItemsController (or ItemCustomersController.. doesn't matter). If you route that form to "customer_items#create", that is the most RESTful option to indicate that you're creating a joins record. 

This would also remove the need to check if the request is a post, but nice work setting up this conditional logic!! At the end of the `create` action in the CustomerItemsController, you can redirect the request to "/customers/:id" to get back to the show page

File: app/models/customer.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class Customer < ApplicationRecord
+  has_and_belongs_to_many :items
```
Comment (juliet-e) [inline app/models/customer.rb:2]: What we've taught you so far with many-to-many-tables doesn't always work perfectly with this `has_and_belongs_to_many` association. That association is the same _idea_ but if you use that association, your joins table can't have any additional fields besides the foreign key. We want the flexibility to be able to add columns to our joins tables as needed (for example, adding an additional status column to the PetApplications table in the last project). So, we teach students the `has_many :items, through :customer_items` instead of this association so that you have more flexibility. 

tl;dr: while this association can be handy in some cases, use the association `has_many .. through: ` instead because it's more flexible.

File: app/models/customer.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+
+  def add_item(item_id)
```
Comment (juliet-e) [inline app/models/customer.rb:4]: Nice encapsulation of this logic at the model level! This is just .find and << which is all CRUD operations so it's totally fine to leave them in the controller, but I think it cleans things up nicely to have them in their own method!

File: app/models/item.rb [hunk 4]
```diff
@@ -1,3 +1,4 @@
 class Item < ApplicationRecord
   belongs_to :supermarket
+  has_and_belongs_to_many :customers
```
Comment (juliet-e) [inline app/models/item.rb:3]: See note about this in your Customer model!

File: app/views/customers/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  <% @customer.items.each do |item| %>
+    <li>
+      <p>Item: <%= item.name %></p>
+      <p>Price: <%= "$#{item.price}" %></p>
+      Supermarket: <%= item.supermarket.name %>
+    </li>
+  <% end %>
+</ul>
+
+<%= form_with url: "/customers/#{@customer.id}", method: :post, data: {turbo: false} do |form| %>
```
Comment (juliet-e) [inline app/views/customers/show.html.erb:14]: Building off the comment I left in your controller:
the more RESTful path for this form would be "/customers/:id/items"

File: config/routes.rb [hunk 6]
```diff
@@ -3,4 +3,8 @@
 
   # Defines the root path route ("/")
   # root "articles#index"
+
+  get "/customers/:id", to: "customers#show"
+  post "/customers/:id", to: "customers#show"
```
Comment (juliet-e) [inline config/routes.rb:8]: The more RESTful option here would be 
`post "/customers/:customer_id/items", to: "customer_items#create"`

Remember that from that action you can redirect back to the show page to display the same page with the newly added item!
